File: applications/rtkextractshroudsignal/rtkextractshroudsignal.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkextractshroudsignal_ggo.h"
#include "rtkMacro.h"

#include "rtkDPExtractShroudSignalImageFilter.h"
#include "rtkReg1DExtractShroudSignalImageFilter.h"
#include "rtkExtractPhaseImageFilter.h"

#include <itkImageFileReader.h>
#include <fstream>

namespace rtk
{
template <class TSignalType>
void
WriteSignalToTextFile(TSignalType * sig, const std::string & fileName)
{
  std::ofstream                              ofs(fileName.c_str());
  itk::ImageRegionConstIterator<TSignalType> it(sig, sig->GetLargestPossibleRegion());
  for (it.GoToBegin(); !it.IsAtEnd(); ++it)
  {
    ofs << it.Get() << std::endl;
  }
  ofs.close();
}
} // namespace rtk

int
main(int argc, char * argv[])
{
  GGO(rtkextractshroudsignal, args_info);

  using InputPixelType = double;
  using OutputPixelType = double;
  constexpr unsigned int Dimension = 2;

  using InputImageType = itk::Image<InputPixelType, Dimension>;
  using OutputImageType = itk::Image<OutputPixelType, Dimension - 1>;

  // Extract shroud signal
  OutputImageType::Pointer shroudSignal;
  if (std::string(args_info.method_arg) == "DynamicProgramming")
  {
    if (!args_info.amplitude_given)
    {
      std::cerr << "You must supply a maximum amplitude to look for." << std::endl;
      return 1;
    }
    using shroudFilterType = rtk::DPExtractShroudSignalImageFilter<InputPixelType, OutputPixelType>;
    auto shroudFilter = shroudFilterType::New();
    shroudFilter->SetInput(itk::ReadImage<InputImageType>(args_info.input_arg));
    shroudFilter->SetAmplitude(args_info.amplitude_arg);
    TRY_AND_EXIT_ON_ITK_EXCEPTION(shroudFilter->Update())
    shroudSignal = shroudFilter->GetOutput();
  }
  else if (std::string(args_info.method_arg) == "Reg1D")
  {
    using shroudFilterType = rtk::Reg1DExtractShroudSignalImageFilter<InputPixelType, OutputPixelType>;
    auto shroudFilter = shroudFilterType::New();
    shroudFilter->SetInput(itk::ReadImage<InputImageType>(args_info.input_arg));
    TRY_AND_EXIT_ON_ITK_EXCEPTION(shroudFilter->Update())
    shroudSignal = shroudFilter->GetOutput();
  }
  else
  {
    std::cerr << "The specified method does not exist." << std::endl;
    return 1;
  }

  // Write output signal
  rtk::WriteSignalToTextFile(shroudSignal.GetPointer(), args_info.output_arg);

  // Process phase signal if required
  if (args_info.phase_given)
  {
    using PhaseFilter = rtk::ExtractPhaseImageFilter<OutputImageType>;
    auto phase = PhaseFilter::New();
    phase->SetInput(shroudSignal);
    phase->SetMovingAverageSize(args_info.movavg_arg);
    phase->SetUnsharpMaskSize(args_info.unsharp_arg);
    phase->SetModel((PhaseFilter::ModelType)args_info.model_arg);
    TRY_AND_EXIT_ON_ITK_EXCEPTION(phase->Update())

    rtk::WriteSignalToTextFile(phase->GetOutput(), args_info.phase_arg);
  }

  return EXIT_SUCCESS;
}

File: applications/rtkextractshroudsignal/rtkextractshroudsignal.py
#!/usr/bin/env python
import argparse
from cmath import phase
import sys
import itk
from itk import RTK as rtk


def write_signal_to_text_file(signal_image, filename):
    # Convert ITK image to numpy array
    signal_array = itk.array_from_image(signal_image)

    # Write to text file
    with open(filename, 'w') as f:
        for value in signal_array.flatten():
            f.write(f"{value}\n")


def build_parser():
    parser = argparse.ArgumentParser(
        description="Extracts the breathing signal from a shroud image.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )

    # General options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--input", "-i", help="Input shroud image file name", type=str, required=True
    )
    parser.add_argument(
        "--amplitude", "-a", help="Maximum breathing amplitude explored in mm", type=float
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--method", "-m",
        help="Method to use",
        choices=["Reg1D", "DynamicProgramming"],
        default="Reg1D"
    )

    # Phase extraction options
    phase_group = parser.add_argument_group("Phase extraction")
    phase_group.add_argument(
        "--phase", "-p", help="Output file name for the Hilbert phase signal", type=str
    )
    phase_group.add_argument(
        "--movavg", help="Moving average size applied before phase extraction", type=int, default=1
    )
    phase_group.add_argument(
        "--unsharp", help="Unsharp mask size applied before phase extraction", type=int, default=55
    )
    phase_group.add_argument(
        "--model",
        help="Phase model",
        type=str,
        choices=["LOCAL_PHASE", "LINEAR_BETWEEN_MINIMA", "LINEAR_BETWEEN_MAXIMA"],
        default="LINEAR_BETWEEN_MINIMA"
    )

    return parser


def process(args_info: argparse.Namespace):
    InputPixelType = itk.D
    OutputPixelType = itk.D
    Dimension = 2

    OutputImageType = itk.Image[OutputPixelType, Dimension - 1]

    # Read input image
    if args_info.verbose:
        print(f"Reading input shroud image from {args_info.input}...")

    input_image = itk.imread(args_info.input, InputPixelType)

    # Extract shroud signal
    if args_info.verbose:
        print(f"Extracting shroud signal using method: {args_info.method}")

    if args_info.method == "DynamicProgramming":
        if args_info.amplitude is None:
            print("Error: You must supply a maximum amplitude to look for when using DynamicProgramming method.")
            sys.exit(1)

        shroud_filter = rtk.DPExtractShroudSignalImageFilter[InputPixelType, OutputPixelType].New()
        shroud_filter.SetInput(input_image)
        shroud_filter.SetAmplitude(args_info.amplitude)

        shroud_filter.Update()
        shroud_signal = shroud_filter.GetOutput()

    elif args_info.method == "Reg1D":
        shroud_filter = rtk.Reg1DExtractShroudSignalImageFilter[InputPixelType, OutputPixelType].New()
        shroud_filter.SetInput(input_image)

        shroud_filter.Update()
        shroud_signal = shroud_filter.GetOutput()

    else:
        print(f"Error: The specified method '{args_info.method}' does not exist.")
        sys.exit(1)

    # Write output signal
    if args_info.verbose:
        print(f"Writing shroud signal to {args_info.output}...")

    write_signal_to_text_file(shroud_signal, args_info.output)

    # Process phase signal if requested
    if args_info.phase:
        phase = rtk.ExtractPhaseImageFilter[OutputImageType].New()
        phase.SetInput(shroud_signal)
        phase.SetMovingAverageSize(args_info.movavg)
        phase.SetUnsharpMaskSize(args_info.unsharp)
        # Map string to enum integer value
        model_values = {
            "LOCAL_PHASE": 0,
            "LINEAR_BETWEEN_MINIMA": 1,
            "LINEAR_BETWEEN_MAXIMA": 2
        }
        phase.SetModel(model_values[args_info.model])

        phase.Update()

        if args_info.verbose:
            print(f"Writing phase signal to {args_info.phase}...")

        write_signal_to_text_file(phase.GetOutput(), args_info.phase)

    if args_info.verbose:
        print("Processing completed successfully.")


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
